### PR TITLE
Implement travel time component

### PR DESCRIPTION
**Pull request by DanielHeath**

### CODE REVIEW

File: src/components/StaticMap.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ class Address extends React.Component {
     };
 
     render(): ReactElement {
-        const baseUrl = "//maps.googleapis.com/maps/api/staticmap";
+        const baseUrl = "https://maps.googleapis.com/maps/api/staticmap";
```
Comment (danni) [inline src/components/StaticMap.js:29]: Looks suspect. Won't mixing http and https URLs generate a browser warning?

File: src/components/TransportTime.js [hunk 2]
```diff
@@ -7,6 +6,9 @@ import { titleize } from "underscore.string";
 import fixtures from "../../fixtures/services";
 import icons from "../icons";
 import Location from "../iss/Location";
+import storage from "../storage";
+
+declare var google: Google;
```
Comment (danni) [inline src/components/TransportTime.js]: Use the Google Maps promise instead.

File: src/components/TransportTime.js [hunk 3]
```diff
@@ -29,12 +31,52 @@ class TransportTime extends React.Component {
...
+
+        const options = {
+            origin: `${location}`,
+            destination: `${destPoint.lat},${destPoint.lon}`,
+            travelMode: google.maps.DirectionsTravelMode.WALKING,
+            unitSystem: google.maps.UnitSystem.METRIC,
+            language: "en-AU",
+        };
+
+        let directionsService = new google.maps.DirectionsService();
```
Comment (danni) [inline src/components/TransportTime.js]: `const Maps = await Maps();`

File: src/components/TransportTime.js [hunk 4]
```diff
@@ -29,12 +31,52 @@ class TransportTime extends React.Component {
...
+        if (coords && coords.latitude && coords.longitude) {
+            location = `${coords.latitude},${coords.longitude}`;
+        }
+        // FIXME: check google.maps.DirectionsTravelMode.TRANSIT
+        // once we expand beyond VIC (or transit is implemented in VIC).
+
+        const options = {
+            origin: `${location}`,
+            destination: `${destPoint.lat},${destPoint.lon}`,
+            travelMode: google.maps.DirectionsTravelMode.WALKING,
```
Comment (danni) [inline src/components/TransportTime.js]: Public transport if > 2km? It's only Melbourne where this doesn't work.

File: src/components/TransportTime.js [hunk 5]
```diff
@@ -29,12 +31,52 @@ class TransportTime extends React.Component {
...
+
+        let directionsService = new google.maps.DirectionsService();
+
+        directionsService.route(options, (response, status) => {
+            if (status == google.maps.DirectionsStatus.OK) {
+                this.setState({time: response.routes[0].legs.map(
+                    (leg) => leg.duration.text
+                ).reduce(
+                    (memo, text) => `${memo} then ${text}`
+                )});
```
Comment (danni) [inline src/components/TransportTime.js]: Can we just use the total time?

File: src/components/StaticMap.js (hunk 1 above)
Comment (DanielHeath) [inline src/components/StaticMap.js:29]: Fetching secure resources from an insecure page doesn't generate a warning (only vice versa)

File: src/server/HtmlDocument.js [hunk 6]
```diff
@@ -235,6 +235,9 @@ class HtmlDocument extends React.Component {
             dangerouslySetInnerHTML={{__html: markup}}
         />
         <script dangerouslySetInnerHTML={{__html: envConfig}} />
+
+        <script src={gmapsApi}>
+        </script>
```
Comment (danni) [inline src/server/HtmlDocument.js]: Why?

File: src/components/TransportTime.js [hunk 7]
```diff
@@ -29,12 +31,52 @@ class TransportTime extends React.Component {
...
+            origin: `${location}`,
+            destination: `${destPoint.lat},${destPoint.lon}`,
+            travelMode: google.maps.DirectionsTravelMode.WALKING,
+            unitSystem: google.maps.UnitSystem.METRIC,
+            language: "en-AU",
+        };
+
+        let directionsService = new google.maps.DirectionsService();
+
+        directionsService.route(options, (response, status) => {
```
Comment (danni) [inline src/components/TransportTime.js]: The other promise wrappers are in `src/maps.js` let's be consistent.

File: src/components/TransportTime.js (hunk 3 above)
Comment (DanielHeath) [inline src/components/TransportTime.js]: where does `Maps()` come from?

Comment (DanielHeath) [inline src/components/TransportTime.js]: ooh I see. Nice :)

File: src/components/TransportTime.js (hunk 4 above)
Comment (DanielHeath) [inline src/components/TransportTime.js]: Sure, but that's where we're launching in a few weeks.

File: src/server/HtmlDocument.js (hunk 6 above)
Comment (DanielHeath) [inline src/server/HtmlDocument.js]: No longer needed with the maps promise stuff.

File: src/components/TransportTime.js (hunk 4 above)
Comment (danni) [inline src/components/TransportTime.js]: We're actually launching nationwide IIRC.

File: src/components/TransportTime.js (hunk 5 above)
Comment (DanielHeath) [inline src/components/TransportTime.js]: We could sum the time numbers and then humanize them (moment.js does it too fuzzily imo - 45 mins turns into an hour). There's no total in the returned object.

Comment (danni) [inline src/components/TransportTime.js]: Probably doesn't even need humanising? Also if we get the cost field, should we include that?

Comment (DanielHeath) [inline src/components/TransportTime.js]: I guess we could just say 'X minutes'

**Comment by DanielHeath** [comment]
@danni ++?

**Comment by danni** [comment]
This is super over complex, can we just put a request in following the ISS request and add it to the Location object? Then we could cache the whole lot at once and just get the travel time from the location object.

**Comment by DanielHeath** [comment]
That's probably much simpler, I'll give it a shot.

File: src/iss.js [hunk 8]
```diff
@@ -151,16 +152,48 @@ export async function request(
...
+): Promise<Array<Service>> {
+    let formatPoint = (point) => `${point.lat},${point.lon}`;
+    const maps = await Maps();
+    let service: ?Service;
+    let travelTimes = await maps.travelTime(
+        [
+            /*::`*/
+            for (service of services)
+                if (!service.Location().isConfidential())
+                    formatPoint(service.location.point)
```
Comment (danni) [inline src/iss.js]: Weird indenting for comprehension.

File: src/iss.js [hunk 9]
```diff
@@ -205,22 +238,8 @@ export class Service {
...
-        postcode: string,
-        state: state,
-        street_name: string,
-        street_number: string,
-        street_suffix: string,
-        street_type: string,
-        suburb: string,
-    };
+    location: issLocation;
+    travelTime: ?travelTime;
```
Comment (danni) [inline src/iss.js]: Comment explaining this is from Google please :)

File: src/maps.js [hunk 10]
```diff
@@ -9,6 +12,84 @@ class MapsApi {
...
+                    }
+                }
+            }
+        )
+    }
+
+    batchDirectionsRequest(
+        destinations: Array<string>, mode = "WALKING"
+    ): Promise<Array<travelTime>> {
+        let directionsService = new this.api.DistanceMatrixService();
```
Comment (danni) [inline src/maps.js]: const ?

File: src/maps.js [hunk 11]
```diff
@@ -9,6 +12,84 @@ class MapsApi {
...
+        const coords = storage.getJSON("coordinates");
+        let origin = storage.getItem("location");
+
+        if (coords && coords.latitude && coords.longitude) {
+            origin = `${coords.latitude},${coords.longitude}`;
+        }
+
+        let params = {
```
Comment (danni) [inline src/maps.js]: const ?

File: src/maps.js [hunk 12]
```diff
@@ -61,6 +142,12 @@ function maps(): Promise<MapsApi> {
         function checkLoaded() {
             try {
                 google.maps;
+                if (!google.maps.DistanceMatrixService) {
+                    throw new Error(
+                        "DistanceMatrixService comes in late",
+                    );
+                }
```
Comment (danni) [inline src/maps.js]: I don't understand what this is?

File: test/features/10-functional/10-basic/22-transport-times.feature [hunk 13]
```diff
@@ -14,17 +14,18 @@ Feature: Transport times
...
+        location:
+            suburb: Richmond
+            point:
+                lat: -37.8578706
+                lon: 144.5192432
         ----------------------------------------------
-
         When I search for the service
-        Then I should see a transport time of "? mins Richmond"
+        Then I should see a transport time of "9 hours 21 mins Richmond"
```
Comment (DanielHeath) [inline test/features/10-functional/10-basic/22-transport-times.feature]: This needs to be stubbed or it'll get random flaky results

Comment (DanielHeath) [inline test/features/10-functional/10-basic/22-transport-times.feature]: And the icon should have a screenreader tag

File: src/maps.js (hunk 12 above)
Comment (DanielHeath) [inline src/maps.js]: That's for flow

Comment (DanielHeath) [inline src/maps.js]: replace this with a reference to `google.maps.DistanceMatrixService.<method>`

File: src/iss.js [hunk 14]
```diff
@@ -151,16 +152,46 @@ export async function request(
...
+): Promise<Array<Service>> {
+    let formatPoint = (point) => `${point.lat},${point.lon}`;
+    const maps = await Maps();
+    let service: ?Service;
+    let travelTimes = await maps.travelTime(
+        [/*::`*/
+        for (service of services)
+            if (!service.Location().isConfidential())
+                formatPoint(service.location.point)
+        /*::`*/]
```
Comment (danni) [inline src/iss.js]: Still has weird indenting. I expect the three rows to be flat. Consider repatriating `[]` with `()`

**Comment by danni** [comment]
++

**Pull request merged**